File: client/tws_foundation_client/lib/src/services/models/inputs/view_input.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

/// {model} class for [ViewInput].
///
/// Defines a data model object that represents a {View} operation {input} object.
final class ViewInput<T extends EntityI<T>> implements EncodableI {
  /// Last created time, this works to limit the new entries on demand with [retroactive].
  DateTime? timestamp;

  /// Whether the {View} engine should consider records created after the [timestamp] last generated [ViewOutput].
  bool retroactive = false;

  /// Whether the {View} operation is for an exportable result context.
  bool export = false;

  /// Amount of items per [page].
  int range = 0;

  /// Specific page to get items.
  int page = 0;

  /// Filtering instructions for [EntityI] properties matching delegates. They are applied by the given [ViewFilterNodeI.order] property.
  List<ViewFilterNodeI<T>> filters = <ViewFilterI<T>>[];

  /// Ordering instructions per [EntityI] property, they are applied by the collection order and are not override, meaning
  /// all orderings are applied at the same time.
  List<ViewOrdering> orderings = <ViewOrdering>[];

  /// Creates a new [ViewInput] instance.
  ViewInput();

  /// Creates a new [ViewInput] instance based on relevant properties.
  ViewInput.a(this.range, this.page, this.orderings, this.filters);

  /// Creates a new [ViewInput] instance based on specific paging properties.
  ViewInput.b(this.range, this.page);

  @override
  DataMap encode() {
    return <String, Object?>{
      EntityKeys.timestamp: timestamp,
      'retroactive': retroactive,
      'export': export,
      'range': range,
      'page': page,
      'filters': filters.encode(),
      'orderings': orderings.encode(),
    };
  }
}

/// {model} class for [ViewOrdering].
///
/// Deefines a data model object that represents an [EntityI] property ordering instruction to the
/// {View} generation engine.
final class ViewOrdering implements EncodableI {
  /// Name of the [EntityI] property to be ordered.
  String property = '';

  /// Ordering instruction.
  ViewOrderings ordering = ViewOrderings.ascending;

  /// Creates a new [ViewOrdering] instance.
  ViewOrdering();

  @override
  DataMap encode() {
    return <String, Object?>{
      'property': property,
      'ordering': ordering.index,
    };
  }
}

/// {enum} implementation for [ViewOrderings].
///
/// Stores the availbale ordering options for [ViewOrdering.ordering] property.
enum ViewOrderings {
  /// Whether the [ViewOrdering.property] must be ordered by ascending.
  ascending,

  /// Whether the [ViewOrdering.property] must be ordered by descending.
  descending,
}

File: server/TWS Foundation/CSM Foundation/Database/Entity/Depot/IDepot_View/ViewInput.cs
using CSM_Foundation.Database.Entity.Depot.IDepot_View.ViewFilters;
using CSM_Foundation.Database.Entity.Models.Input;

namespace CSM_Foundation.Database.Entity.Depot.IDepot_View;

/// <summary>
///     Defines options to build a [View]
///     specifing the behavior to the builder.
/// </summary>
public class ViewInput<T>
    where T : IEntity {

    /// <summary>
    ///     On <see langword="true"/> indicate that the builder should consider all the new items added 
    ///     after the <see cref="Timestamp"/> if it is null then won't consider the limitation will behave
    ///     as this property is <see langword="true"/>
    /// </summary>
    public required bool Retroactive { get; init; }

    /// <summary>
    ///     Specifies the amount of items expected per page
    /// </summary>
    public required int Range { get; init; }

    /// <summary>
    ///     Specifies the current desired page.
    /// </summary>
    public required int Page { get; init; }

    /// <summary>
    ///     Specifies the last time this view was created, this works to limit the new entries 
    ///     on demand by <see cref="Retroactive"/>
    /// </summary>
    public DateTime? Timestamp { get; init; }

    /// <summary>
    ///     Indicates if the current options are to generate an [Export] result.
    /// </summary>
    public bool Export { get; set; } = false;

    /// <summary>
    ///     Indicates order actions to perform to the current view building.
    ///     <br>
    ///         The ordering actions will be performed by the array received order.
    ///     </br>
    /// </summary>
    public ViewOrdering[] Orderings { get; init; } = [];

    /// <summary>
    ///     How the [View] data will be filtered.
    /// </summary>
    public IViewFilterNode<T>[] Filters { get; init; } = [];
}
